design/one-pager-stack-ui-metadata.md

```diff
@@ -0,0 +1,170 @@
+  - name: dbReplicas
+    controlType: singleInput
+    type: integer
+    path: .spec.dbReplicas
+    title: DB Replicas
+    description: The number of DB Replicas
+    default: 1
+    validation:
+    - minimum: 1
+    - maximum: 3
+  - name: masterPassword
+    controlType: singleInput
+    type: password
+    path: .spec.masterPassword
+    title: DB Master Password
+    description: The master DB password. Must be between 8-32 characters long
+  - name: subdomain
+    controlType: singleInput
+    type: string
+    path: .spec.subdomain
+    title: Subdomain
+    pattern: ^([A-Za-z0-9](?:(?:[-A-Za-z0-9]){0,61}[A-Za-z0-9])?){2,62}$
+    description: Enter a value for your subdomain. It cannot start or end with a dash and must be between 2-62 characters long
+    validation:
+    - minLength: 2
+    - maxLength: 62
+  - name: instanceSize
+    controlType: singleSelect
+    path: .spec.instanceSize
+    title: Instance Size
+    enum:
+    - 'Option-1'
+    - 'Option-2'
+    - 'Option-3'
+    validation:
+    - required: true
+      customError: You must select an instance size for your configuration!
+```
+
+### CRD Annotation Example
+
+An example injection of the `ui-schema.yaml` as a CRD annotation follows.  Keep in mind that Kubernetes annotations are limited to 256kb (less in older versions).
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    extensions.crossplane.io/ui-spec: {"uiSpecVersion":0.3,"uiSpec":[{"title":"Configuration","description":"Enter information specific to the configuration you wish to create.","items":[{"name":"dbReplicas","controlType":"singleInput","type":"integer","path":".spec.dbReplicas","title":"DB Replicas","description":"The number of DB Replicas","default":1,"validation":[{"minimum":1},{"maximum":3}]},{"name":"masterPassword","controlType":"singleInput","type":"password","path":".spec.masterPassword","title":"DB Master Password","description":"The master DB password. Must be between 8-32 characters long"},{"name":"subdomain","controlType":"singleInput","type":"string","path":".spec.subdomain","title":"Subdomain","pattern":"^([A-Za-z0-9](?:(?:[-A-Za-z0-9]){0,61}[A-Za-z0-9])?){2,62}$","description":"Enter a value for your subdomain. It cannot start or end with a dash and must be between 2-62 characters long","validation":[{"minLength":2},{"maxLength":62}]},{"name":"instanceSize","controlType":"singleSelect","path":".spec.instanceSize","title":"Instance Size","enum":["Option-1","Option-2","Option-3"],"validation":[{"required":true,"customError":"You must select an instance size for your configuration!"}]}]}]}
+  creationTimestamp: null
+  labels:
+    controller-tools.k8s.io: "1.0"
+  name: rdsinstances.database.aws.crossplane.io
+.
+.
+.
+.
+.
+```
+
+## Formal Specification
+
+The current design calls for unbiased processing of the UI YAML to annotations, regardless of content.
+
+Because the YAML will be transcribed as JSON in the annotation:
```

Comment:
something i hadn't really thought about until now: the stack package author writes the UI markup content in yaml, but transcribe (re-encode) into JSON to store as an annotation on the CRD?

Does it make more sense to expect the stack author to initially write their UI markup in JSON, OR to store the UI markup in the annotation as YAML?

Reply:
JSON is valid YAML: https://en.wikipedia.org/wiki/YAML#Comparison_with_JSON (some restrictions apply)

For hand-editing, I think escaping YAML within YAML may be more intricate that doing so with JSON. 
For tool digestion, I don't think it matters if the expression of UI in the Kubernetes object is JSON or YAML.
For reviewing Kubernetes objects and their properties by sight, I might prefer YAML.

If we use a YAML parser instead of a JSON parser when interpreting these fields we can accommodate both.  I have what is probably a misconception that nested or embedded YAML property values have some problems.  Other than the sense of some forgotten readings, my only evidence is that JSON is often embedded.

Let's aim for YAML.

---

@suskin asked yesterday why we even need an `app.yaml`  and `ExtensionRequest` when the `Extension` (and its `status`) could serve as both.  Working on inertia, I couldn't offer a reason.

Reply:
Let's revisit this in #604